### PR TITLE
Feature resend decide ss

### LINKED ISSUE #68: Complete reconnection automation
Together issue #61  and issue #67 enable us to discover and react to dropped `Prepare`, `AcceptSync`, `AcceptDecide`, and `Decide` messages rather than relying on the user to call reconnected().

They don't cover cases of dropped `AcceptStopSign`, `DecideStopSign`, `PrepareReq` messages. The absence of these messages can't be easily detected with sequence numbers. Instead they can be resent after some timeout (can for example use the same timeout as BLE).

### PR DESCRIPTION
**Pull request by kevin-harrison**
Please make sure these boxes are checked, before submitting a new PR.

- [x] You ran the local CI checker with `./check.sh` with no errors
- [x] You reference which issue is being closed in the PR text (if applicable)
- [ ] You updated the OmniPaxos book (if applicable)

## Issues

Fix #68  by resending the decides of StopSigns.

## Breaking Changes
 - The trait function `Storage::set_stopsign(&mut self, s: StopSign) -> StorageResult<()>`'s signature changes to `Storage::set_stopsign(&mut self, s: Option<StopSign>) -> StorageResult<()>`.

## Other Changes
 - Fixed AcceptSync bugs by changing how Stopsign's and Snapshots are created.
 - Accounts for a bug in CommitLog when appending an empty set of entries.

### CODE REVIEW

**Review by haraldng — COMMENTED** [review]
I think the code looks mostly correct (please check comments to see if suggested changes make sense). 

I just want to understand the changes related to the snapshot better: How/when do we create an empty snapshot? and how does it break the decided index?

File: omnipaxos/src/sequence_paxos/mod.rs [hunk 1]
```diff
@@ -214,20 +214,21 @@ where
...
-                        if !ss.decided(decided_idx) {
-                            for follower in self.leader_state.get_promised_followers() {
-                                // resend if the follower has not accepted the stopsign
-                                if !ss.decided(self.leader_state.get_accepted_idx(follower)) {
-                                    self.send_accept_stopsign(follower, ss.stopsign.clone(), true);
-                                }
+
+                        for follower in self.leader_state.get_promised_followers() {
+                            if !ss.decided(decided_idx)
+                                && !ss.decided(self.leader_state.get_accepted_idx(follower))
```
[the remark this answers is not in the document]
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/mod.rs:224]: Not 100% sure what you mean by losing last N accepts. But i think either the next message received is then either a prepare (we will resync and get missed accepts) or an acceptDecide/decide/acceptStopSign and the sequence break will be discovered.

[the remark this answers is not in the document]
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/mod.rs:224]: When we create an empty stopsign:
- node A disconnects from cluster
- cluster proposes and decides a SS
- node reconnects to cluster
- leader creates AccSync with suffix=[ ], decided_idx=1 a stopsign and because the leader sees the it has a higher decided idx, it creates a snapshot of the empty log (can't snapshot a SS).
- on receiving the AccSync now Node A has a messed up log because it takes the SS and also the snapshot representing and empty log

[the remark this answers is not in the document]
Comment (haraldng) [reply omnipaxos/src/sequence_paxos/mod.rs]: > Not 100% sure what you mean by losing last N accepts. But i think either the next message received is then either a prepare (we will resync and get missed accepts) or an acceptDecide/decide/acceptStopSign (the sequence break will be discovered).

Consider we send N Accept messages and then send a StopSign and they all get dropped. Since a StopSign is pending, the leader will not propose any further entries. Therefore we will never decide any of these unless we periodically send some message so that we can detect something has been dropped.

[the remark this answers is not in the document]
Comment (haraldng) [reply omnipaxos/src/sequence_paxos/mod.rs]: but maybe we can solve that in a future PR

[the remark this answers is not in the document]
Comment (haraldng) [reply omnipaxos/src/sequence_paxos/mod.rs]: on second thought, this is fine with your changes of resending stopsigns. But let's say we don't propose SS and don't propose any new entries after those N entries. Then those N entries will not be decided

File: omnipaxos/src/messages.rs [hunk 2]
```diff
@@ -73,6 +73,8 @@ pub mod sequence_paxos {
         pub sync_idx: u64,
         /// The decided index
         pub decided_idx: u64,
+        /// The accepted index
+        pub accepted_idx: u64,
```
Comment (haraldng) [inline omnipaxos/src/messages.rs:77]: Why is this needed? Isn't it sync_idx + suffix.len() + 0 or 1 (if StopSign is some)? I.e., we can calculate this by calling `self.internal_storage.get_accepted_idx()` at the receiver.

It is probably equally correct to explicitly send the accepted_idx. But I prefer if we use the `self.internal_storage.get_accepted_idx()` for consistency (and avoid any potential bugs)

File: omnipaxos/src/sequence_paxos/follower.rs [hunk 3]
```diff
@@ -119,12 +119,19 @@ where
                             "storage error while trying to read old log or snapshot",
                         );
                     }
+                    // Compact index of snapshot depends on if incoming stopsign is decided
+                    let new_compact_idx = match accsync.stopsign {
+                        Some(_) if accsync.decided_idx == accsync.accepted_idx => {
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/follower.rs:124]:
```suggestion
                        Some(ss) if accsync.decided_idx == ss.log_idx => {
```

File: omnipaxos/src/messages.rs [hunk 4]
```diff
@@ -73,6 +73,8 @@ pub mod sequence_paxos {
         pub sync_idx: u64,
```
Comment (haraldng) [inline omnipaxos/src/messages.rs:73]: To avoid confusion I think we should replace the `decided_snapshot, suffix, sync_idx` fields with the following:
```
enum SyncItem {
        Entries(u64, Vec<T>), // sync_idx + suffix (receiver applies suffix at sync_idx)
        SnapshotAndEntries(SnapshotType<T>, Vec<T>) // receiver applies snapshot (with decided_idx) and appends suffix)
}

```

File: omnipaxos/src/sequence_paxos/follower.rs [hunk 5]
```diff
@@ -167,26 +169,19 @@ where
...
-                        }
-                    } else {
-                        self.accept_stopsign(ss);
-                    }
-                    accepted.accepted_idx = self.internal_storage.get_accepted_idx();
-                }
-                None => self.forward_pending_proposals(),
+            if accsync.stopsign.is_none() {
+                self.forward_pending_proposals();
+                //TODO: foward pending stopsign like in handle_majority_promises?
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/follower.rs:176]: no need to forward StopSign here. In handle_majority_promises we need to adopt the StopSign because it might be chosen. Here it is not chosen so don't need to do it.

File: omnipaxos/src/sequence_paxos/mod.rs [hunk 6]
```diff
@@ -214,23 +214,22 @@ where
...
+                        for follower in self.leader_state.get_promised_followers() {
+                            let decided_idx = self.internal_storage.get_decided_idx();
+                            if self.leader_state.get_accepted_idx(follower)
+                                != self.internal_storage.get_accepted_idx()
+                            {
+                                self.send_accept_stopsign(follower, ss.clone(), true);
+                            }
+                            if self.internal_storage.stopsign_is_decided() {
+                                self.send_decide(follower, decided_idx, true);
                             }
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/mod.rs:229]: are the send_accept_stopsign and send_decide here in the wrong order (i.e., the sends are inside the wrong if)?

File: omnipaxos/src/sequence_paxos/mod.rs [hunk 7]
```diff
@@ -214,23 +214,22 @@ where
...
-                        if !ss.decided(decided_idx) {
-                            for follower in self.leader_state.get_promised_followers() {
-                                // resend if the follower has not accepted the stopsign
-                                if !ss.decided(self.leader_state.get_accepted_idx(follower)) {
-                                    self.send_accept_stopsign(follower, ss.stopsign.clone(), true);
-                                }
+                        for follower in self.leader_state.get_promised_followers() {
+                            let decided_idx = self.internal_storage.get_decided_idx();
+                            if self.leader_state.get_accepted_idx(follower)
+                                != self.internal_storage.get_accepted_idx()
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/mod.rs:223]: Just to understand what we're doing here. Say if the leader has accepted_idx = 10 and decided_idx = 8, and follower has accepted_idx = 9. Is the idea that by resending `Decide<9>` that the follower will detect a sequence break and then catch up?

**Review by haraldng — CHANGES_REQUESTED** [review]
Please check comments and questions. Also, if I understand correctly, we remove the log index from the StopSign and instead calculate it instead?

File: omnipaxos/src/sequence_paxos/mod.rs [hunk 8]
```diff
@@ -217,16 +217,16 @@ where
...
                 if *phase == Phase::Accept {
                     if let Some(ss) = self.internal_storage.get_stopsign() {
+                        let decided_idx = self.internal_storage.get_decided_idx();
                         for follower in self.leader_state.get_promised_followers() {
-                            let decided_idx = self.internal_storage.get_decided_idx();
-                            if self.leader_state.get_accepted_idx(follower)
+                            if self.internal_storage.stopsign_is_decided() {
+                                self.send_decide(follower, decided_idx, true);
+                            }
+                            else if self.leader_state.get_accepted_idx(follower)
```
Comment (kevin-harrison) [inline omnipaxos/src/sequence_paxos/mod.rs:225]: If we are resending a Decide for a StopSign then our sequence number has progressed past when we sent the AcceptStopSign. Since we don't cache the AcceptStopSign's sequence number if we resent it we would just trigger a resync.

Now if a follower is missing the accept and decide we still trigger a resync when we resend the decide but send less messages (and dont send an AcceptStopSign with a false sequence number).

File: omnipaxos/src/sequence_paxos/follower.rs [hunk 9]
```diff
@@ -167,26 +162,18 @@ where
...
-                    }
-                    accepted.accepted_idx = self.internal_storage.get_accepted_idx();
-                }
-                None => self.forward_pending_proposals(),
+            if accsync.stopsign.is_none() {
+                self.forward_pending_proposals();
             }
+            self.internal_storage
+                .set_stopsign(accsync.stopsign)
+                .expect("storage error while trying to write stopsign");
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/follower.rs:172]: somewhere in this function, we need to set the decided_idx set by the leader.

**Review by haraldng — CHANGES_REQUESTED** [review]

File: omnipaxos/src/sequence_paxos/follower.rs (hunk 9 above)
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/follower.rs:172]: This happens on line 102

File: omnipaxos/src/sequence_paxos/leader.rs [hunk 10]
```diff
@@ -219,43 +219,52 @@ where
...
                         .expect("storage error while trying to read diff snapshot");
                     let suffix = self
                         .internal_storage
                         .get_suffix(my_decided_idx)
                         .expect("storage error while trying to read log suffix");
-                    (Some(delta_snapshot), suffix, follower_decided_idx)
+                    let compacted_idx = match self.internal_storage.stopsign_is_decided() {
+                        true => my_decided_idx - 1,
+                        false => my_decided_idx,
+                    };
```
[the remark this answers is not in the document]
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/leader.rs:238]: The problem is we are creating a snapshot diff for the follower and the leader doesn't adopt this diff themself. Because the leader doesn't adopt the snapshot we can't use the leader's `get_compacted_idx()`,  and instead calculate it (snapshot always goes up to leaders decided (-1 if decided stopsign)).

[the remark this answers is not in the document]
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/leader.rs:238]: If the leader would adopt the snapshot here it would have to create the snashot diff (follower_dec, leader_dec) for the follower, then create and adopt a snapshot (leader_compactIdx, leader_dec), and then we could rely on the leader's `get_compacted_idx()`.

File: omnipaxos/src/messages.rs (hunk 4 above)
Comment (kevin-harrison) [reply omnipaxos/src/messages.rs:73]: I think of the AcceptSync as 3 separate steps (adopt Snapshot, adopt suffix, then adopt StopSign). Trying to keep these updates logically separated will make it easier to define the atomic transaction in the future.

File: omnipaxos/src/storage.rs [hunk 11]
```diff
@@ -693,19 +693,22 @@ where
         }
     }
 
+    fn create_decided_snapshot(&mut self) -> StorageResult<T::Snapshot> {
+        self.create_snapshot(self.get_decided_idx())
```
Comment (kevin-harrison) [inline omnipaxos/src/storage.rs:697]: This is incorrect because we don't take into account if a stopsign is included in the decided entries. It so we need to snapshot 1 less item.

File: omnipaxos/src/sequence_paxos/leader.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (haraldng) [reply omnipaxos/src/sequence_paxos/leader.rs:238]: Right, that's true.

File: omnipaxos/src/messages.rs (hunk 4 above)
Comment (haraldng) [reply omnipaxos/src/messages.rs:73]: Yes, but I think this SyncItem would help with that. Because then when matching the different variants we would potentially rollback what is required. 

I don't like that we send a compacted index (that is 0) in the cases we don't even send a snapshot, and similarly a sync index when we use snapshots.

File: omnipaxos/src/storage.rs [hunk 12]
```diff
@@ -704,14 +712,25 @@ where
...
+            true => self.get_decided_idx() - 1,
+            false => self.get_decided_idx(),
+        };
+        let compacted_idx = self.get_compacted_idx();
+        if from_idx <= compacted_idx {
+            if compacted_idx < log_decided_idx {
+                Ok(Some(SnapshotType::Complete(
+                    self.create_snapshot(log_decided_idx)?,
+                )))
+            } else {
```
Comment (kevin-harrison) [inline omnipaxos/src/storage.rs:726]: This is an optimization because when `compacted_idx == log_decided_idx` our current snapshot is up to date and we can just use it. 

However, it also necessary for when an "empty" snapshot doesn't  act like an identity element, like with `LatestValue` in the tests. This is because if we call `create_snapshot` when our log consists of entirely a snapshot, we will it will merge an empty snapshot onto the current one.

File: omnipaxos/src/storage.rs [hunk 13]
```diff
@@ -693,23 +697,48 @@ where
...
+
     pub(crate) fn get_snapshot(&self) -> StorageResult<Option<T::Snapshot>> {
         self.storage.get_snapshot()
     }
 
+    // Creates a Delta snapshot of entries from `from_idx` to the end of the decided log and also
+    // returns the compacted idx of the created snapshot. If the range of entries contains entries
+    // which have already been compacted a valid delta cannot be created, so creates a Complete
+    // snapshot of the entire decided log instead.
     pub(crate) fn create_diff_snapshot(
```
Comment (haraldng) [inline omnipaxos/src/storage.rs:716]: I'm a bit confused regarding this function. It's called create *diff* but we seem to return `Complete` snapshots?

File: omnipaxos/src/storage.rs [hunk 14]
```diff
@@ -682,6 +682,10 @@ where
 
     pub(crate) fn create_snapshot(&mut self, compact_idx: u64) -> StorageResult<T::Snapshot> {
         let compacted_idx = self.get_compacted_idx();
+        assert!(
+            compact_idx >= compacted_idx,
+            "Can't create snapshot, index {compact_idx} already trimmed"
+        );
```
Comment (haraldng) [inline omnipaxos/src/storage.rs:688]: I prefer if we return a StorageResult Error instead of having this assert.

**Review by haraldng — CHANGES_REQUESTED** [review]

File: omnipaxos/src/storage.rs (hunk 13 above)
Comment (kevin-harrison) [reply omnipaxos/src/storage.rs:716]: Imagine the follower has decided_idx = 2 and the leader decided_idx = 10 but also compacted_idx 5. The follower needs a snapshot representing entries 2-10, but the leader can't create that diff since it has compacted entries 1-5, so the best it can do is merge its own snapshot(1-5) with snap(6-10) = snap(1-10). Its a Complete snapshot because the follower needs to adopt snap(1-10) as is and not merge it onto its own snapshot like it would with a delta.

File: omnipaxos/src/sequence_paxos/leader.rs [hunk 15]
```diff
@@ -205,57 +205,44 @@ where
...
+                    .create_diff_snapshot(followers_decided_idx)
                     .expect("storage error while trying to read diff snapshot");
                 let suffix = self
                     .internal_storage
                     .get_suffix(my_decided_idx)
                     .expect("storage error while trying to read log suffix");
-                (Some(delta_snapshot), suffix, follower_decided_idx)
+                (delta_snapshot, suffix, compacted_idx)
+            } else if followers_promise_n == max_promise_n {
+                // Follower's accepted entries are valid, send any new entries after
```
[the remark this answers is not in the document]
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/leader.rs:233]: Doesn't checking the if with decided_idxs + snapshot first solve this?

There is also the case of followers doing crash recovery where they PrepareRequest the same leader after having crashed for a long time (but they would have `followers_promise_n == my_ballot`), maybe worth adding?

[the remark this answers is not in the document]
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/leader.rs:233]: Ok now I think i understand what you mean, we can avoid the trade offs of having the snapshot check first vs the ballots check first by checking for ballots first and then inside that if checking for snapshots again (similarly to how it was before).

File: omnipaxos/src/storage.rs (hunk 13 above)
Comment (haraldng) [reply omnipaxos/src/storage.rs:716]: Right, makes sense

File: omnipaxos/src/sequence_paxos/leader.rs [hunk 16]
```diff
@@ -217,9 +217,11 @@ where
...
             .get_decided_idx(*pid)
             .expect("Received PromiseMetaData but not found in ld");
-        let (delta_snapshot, suffix, sync_idx) =
-            if followers_decided_idx < my_decided_idx && T::Snapshot::use_snapshots() {
-                // Send missing decided entries in snapshot and accepted entries in suffix
+        let (delta_snapshot, suffix, sync_idx) = if followers_promise_n == max_promise_n {
+            // Follower's accepted entries are valid, send any new entries after
+            if T::Snapshot::use_snapshots() && *followers_accepted_idx < my_decided_idx {
+                // Note: we snapshot from follower's decided and not follower's accepted because
+                // snapshots currently can't handle merging onto accepted entries.
```
Comment (kevin-harrison) [inline omnipaxos/src/sequence_paxos/leader.rs:224]: We don't do the most optimal thing here which would be to send the follower a snapshot of entries (follower_acc_idx, leader_dec_idx) since the followers accepted entries are actually decided the follower just doesn't know it yet. However we would have to send some flag telling the follower to merge onto its accepted entries instead of onto its decided entries (`SnapshotType::Delta(Snapshot, bool)` or something). It makes the snapshot code messier than it already is.

Snapshotting like this is only meaningfully better than the current snapshotting when the follower has received a bunch of accepts from the previous leader but not their corresponding decides. With how sequence numbers work this shouldn't really happen.

File: omnipaxos/src/sequence_paxos/leader.rs [hunk 17]
```diff
@@ -205,58 +205,59 @@ where
...
-                        .internal_storage
-                        .get_suffix(*promise_accepted_idx)
-                        .expect("storage error while trying to read log suffix");
-                    (None, sfx, *promise_accepted_idx)
-                }
-            } else if follower_decided_idx < my_decided_idx && T::Snapshot::use_snapshots() {
-                let delta_snapshot = self
+        let (delta_snapshot, suffix, sync_idx) = if (followers_promise_n == max_promise_n) || (*followers_promise_n == self.leader_state.n_leader) {
+            // Follower's accepted entries are valid, send any new entries after
+            if T::Snapshot::use_snapshots() && *followers_accepted_idx < my_decided_idx {
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/leader.rs:222]: Is the second check really needed? Because we anyway send the diff snapshots from the follower's decided index.

```suggestion
            if T::Snapshot::use_snapshots() {
```

File: omnipaxos/src/sequence_paxos/leader.rs [hunk 18]
```diff
@@ -205,58 +205,59 @@ where
...
                 let suffix = self
                     .internal_storage
                     .get_suffix(my_decided_idx)
                     .expect("storage error while trying to read log suffix");
-                (Some(delta_snapshot), suffix, follower_decided_idx)
+                (delta_snapshot, suffix, compacted_idx)
             } else {
+                let sfx = self
+                    .internal_storage
+                    .get_suffix(*followers_accepted_idx)
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/leader.rs:237]: This is incorrect. With the check of just `if (followers_promise_n == max_promise_n) || (*followers_promise_n == self.leader_state.n_leader)` before here, it becomes incorrect to send from the followers accepted index. Consider the following case: the old leader, call it server A, disconnects with the other servers at ballot `n=1` and with 10 log entries. The other servers elect a new leader at `n=2` and decides 10 new entries (i.e., entry 0-9 were from `n=1`, and 10-19 were from `n=2`). But during the partition, A continued in `n=1` and appended 5 **other** entries (it will append because it thinks it's the leader). Now when it reconnects in `n=2`, it will keep those 5 entries that were added during the partition, which is incorrect (because its log entries from 10-14 will be from `n=1` instead of `n=2`).

Instead, we should send the `max_accepted_idx` (which you removed). It is only up to that point in the log that we know that this follower's log is valid.
```suggestion
                    .get_suffix(max_accepted_idx)
```

File: omnipaxos/src/sequence_paxos/leader.rs [hunk 19]
```diff
@@ -205,58 +205,59 @@ where
...
+                    .expect("storage error while trying to read log suffix");
+                (None, sfx, *followers_accepted_idx)
+            }
+        } else {
+            // Follower's accepted entries are not valid, send everying after follower's decided
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/leader.rs:242]:
```suggestion
            // Follower's accepted entries are not valid. Synchronize by sending a snapshot from the follower's decided index up to leader's decided index and any suffix.
```

**Review by haraldng — CHANGES_REQUESTED** [review]

File: omnipaxos/src/sequence_paxos/leader.rs (hunk 17 above)
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/leader.rs:222]: It is not needed for safety but is an optimization (but we need to use `max_accepted_idx` not `followers_accepted_idx` as you've pointed out). The idea is that if the follower has valid accepted entries >= to the leader's `decided_idx` then we don't have to send them a snapshot of the decided entries. We just need to move the follower's `decided_idx` and potentially send a suffix of new accepted entries, and this is exactly what will happen in the `else` statement.

File: omnipaxos/src/sequence_paxos/leader.rs (hunk 18 above)
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/leader.rs:237]: Yeah you are right. If the follower was synced in the previous round we can only trust up to the `max_accepted_idx` of last round. And if the follower was synced by us then we can trust up to their `accepted_idx`.

File: omnipaxos/src/sequence_paxos/leader.rs (hunk 17 above)
Comment (haraldng) [reply omnipaxos/src/sequence_paxos/leader.rs:222]: Right, makes sense.

File: omnipaxos/src/storage.rs [hunk 20]
```diff
@@ -629,6 +629,13 @@ where
         self.state_cache.decided_idx
     }
 
+    fn get_log_decided_idx(&self) -> u64 {
```
Comment (haraldng) [inline omnipaxos/src/storage.rs:632]:
```suggestion
    fn get_decided_idx_without_stopsign(&self) -> u64 {
```

File: omnipaxos/src/sequence_paxos/leader.rs [hunk 21]
```diff
@@ -204,11 +204,14 @@ where
 
     fn send_accsync(&mut self, to: NodeId) {
         let my_decided_idx = self.get_decided_idx();
+        let my_n = self.leader_state.n_leader;
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/leader.rs:207]:
```suggestion
        let current_n = self.leader_state.n_leader;
```

**Review by haraldng — CHANGES_REQUESTED** [review]
Looks mostly good. Just have a few suggestions for naming. I'll do a detailed pass after renaming.

**Comment by haraldng** [comment]
@kevin-harrison since we are changing when we consider a StopSign decided, we might as well fix a related issue which is that currently, we don't know if a StopSign is decided when we perform `read()`. Perhaps we can add a `decided: bool` flag in the `LogEntry::StopSign` variant [here](https://github.com/haraldng/omnipaxos/blob/master/omnipaxos/src/util.rs#L246).

**Comment by kevin-harrison** [comment]
> @kevin-harrison since we are changing when we consider a StopSign decided, we might as well fix a related issue which is that currently, we don't know if a StopSign is decided when we perform `read()`. Perhaps we can add a `decided: bool` flag in the `LogEntry::StopSign` variant [here](https://github.com/haraldng/omnipaxos/blob/master/omnipaxos/src/util.rs#L246).

Do you prefer changing the enum to `StopSign(StopSign, u64)` and the user has to check with the decided index to determine if its decided. Or do you prefer doing it for the user by the enum into two cases `DecidedStopSign(StopSign)` and `UndecidedStopSign(StopSign)`

File: docs/omnipaxos/reconfiguration.md [hunk 22]
```diff
@@ -21,7 +21,7 @@ let decided_entries: Option<Vec<LogEntry<KeyValue>>> = omnipaxos.  read_decided_
 if let Some(de) = decided_entries {
     for d in de {
         match d {
-            LogEntry::StopSign(stopsign) => {
+            LogEntry::StopSign(stopsign, _stopsign_is_decided) => {
```
Comment (haraldng) [inline docs/omnipaxos/reconfiguration.md:24]: This is correct since we are doing read_decided_suffix before. But for the sake of being clear in the docs, we should emphasize that a new instance should only be started when the stopsign is really decided. So here we should do a check that this flag is true first.

File: omnipaxos/src/util.rs [hunk 23]
```diff
@@ -245,8 +245,9 @@ where
     Trimmed(TrimmedIndex),
     /// The entry has been snapshotted.
     Snapshotted(SnapshottedEntry<T>),
-    /// This Sequence Paxos instance has been stopped for reconfiguration.
-    StopSign(StopSign),
+    /// This Sequence Paxos instance has been stopped for reconfiguration. The accompanying bool
+    /// indicates whether the reconfiguration has been decided or not.
```
Comment (haraldng) [inline omnipaxos/src/util.rs:249]:
```suggestion
    /// This Sequence Paxos instance has been stopped for reconfiguration. The accompanying bool
    /// indicates whether the reconfiguration has been decided or not. If it is `true`, then the OmniPaxos instance for the new configuration can be started.
```

**Review by haraldng — CHANGES_REQUESTED** [review]
Fix minor things related to comments and docs example. I will now do a pass of all changes in the whole PR.

File: omnipaxos/src/sequence_paxos/mod.rs [hunk 24]
```diff
@@ -214,23 +214,21 @@ where
     pub(crate) fn resend_message_timeout(&mut self) {
         match &self.state {
             (Role::Leader, phase) => {
```
Comment (haraldng) [inline omnipaxos/src/sequence_paxos/mod.rs:216]: Do we also need to resend `Prepare` if we are in the prepare phase? I'm thinking in the case where we elect a new leader that sends a prepare to everybody which gets dropped. All servers are still connected so they won't elect a new leader, but we will not make progress because the leader does not complete the prepare phase.

If so then change this code to:
```rust
match &self.state {
     (Role::Leader, Phase::Prepare) => { // resend prepare }
     (Role::Leader, Phase::Accept) => { // your current code }

}
```

File: omnipaxos/tests/atomic_storage_test.rs [hunk 25]
```diff
@@ -526,60 +525,3 @@ fn atomic_storage_majority_promises_test() {
         run_single_test(i);
     }
 }
-
-#[test]
-#[serial]
-fn atomic_storage_majority_accepted_stopsign_test() {
```
Comment (haraldng) [inline omnipaxos/tests/atomic_storage_test.rs:532]: These tests are not needed anymore?

File: omnipaxos/tests/sync_test.rs
Comment (haraldng) [inline omnipaxos/tests/sync_test.rs:1]: Please add small description of each test.

**Review by haraldng — COMMENTED** [review]
Looks good. @kevin-harrison please check comments and modify if needed.

File: omnipaxos/src/sequence_paxos/mod.rs (hunk 24 above)
Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/mod.rs:216]: The code now has the leader resending `Prepare`s  no matter the phase, so long as the `Role` is `Leader`. So this case will be covered. Another question is whether we want the leader to send prepares when their `Phase` is `Recovery` or `None`.

Comment (kevin-harrison) [reply omnipaxos/src/sequence_paxos/mod.rs:216]: Maybe its more clear to explicitly pattern match on the entire state like you suggest?

File: omnipaxos/tests/atomic_storage_test.rs (hunk 25 above)
Comment (kevin-harrison) [reply omnipaxos/tests/atomic_storage_test.rs:532]: The purpose of the test was to check for the atomicity of updating `decided_idx` and the result of the `ss.decided(idx)` function. It was a strange test because if you need to roll back the `decided_idx` you dont have to rollback anything about the stopsign (even when it included the `log_idx` field). But I mainly removed it because the `ss.decided(idx)` function no longer exists.

File: omnipaxos/src/sequence_paxos/mod.rs (hunk 24 above)
Comment (haraldng) [reply omnipaxos/src/sequence_paxos/mod.rs:216]: Right, now I remember. Let's keep it this way, because even in the Accept phase, we might need to send `Prepare` to the followers that haven't promised yet.

**Review by haraldng — APPROVED** [review]
